walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,236 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  /**
+   * Finds a list of potential meeting times based on the meeting request and a list of existing
+   * events and their attendees. Picks time slots if one or more time slots exists so that both
+   * mandatory and optional attendees can attend. Otherwise, returns the time slots that fit just
+   * the mandatory attendees.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    long meetingDurationMinutes = request.getDuration();
+
+    List<TimeRange> mandatoryAttendeesBusyTimeRanges =
+        getAllBusyTimeRanges(events, mandatoryAttendees);
+    List<TimeRange> optionalAttendeesBusyTimeRanges =
+        getAllBusyTimeRanges(events, optionalAttendees);
+
+    sortTimeRangesByStart(mandatoryAttendeesBusyTimeRanges);
+    sortTimeRangesByStart(optionalAttendeesBusyTimeRanges);
+
+    List<TimeRange> mandatoryAttendeesFreeTimeRanges =
+        getFreeTimeRanges(mandatoryAttendeesBusyTimeRanges, meetingDurationMinutes);
+    List<TimeRange> optionalAttendeesFreeTimeRanges =
+        getFreeTimeRanges(optionalAttendeesBusyTimeRanges, meetingDurationMinutes);
+
+    if (mandatoryAttendees.isEmpty()) {
+      return optionalAttendeesFreeTimeRanges;
+    } else if (optionalAttendees.isEmpty()) {
+      return mandatoryAttendeesFreeTimeRanges;
+    } else {
+      return findCommonFreeTimeRanges(
+          mandatoryAttendeesFreeTimeRanges,
+          optionalAttendeesFreeTimeRanges,
+          meetingDurationMinutes);
+    }
+  }
+
+  /**
+   * Constructs all the time ranges that any of the mandatory attendees is busy. These time ranges
+   * cannot be part of the potential meeting time.
+   */
+  private List<TimeRange> getAllBusyTimeRanges(
+      Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+
+      for (String attendee : eventAttendees) {
```

Comment:
I think you could make this a little cleaner and more efficient.

The first thing is that you're calling `.contains(...)` on the `Collection<String> mandatoryAttendees` which could be inefficient if the collection is large, so maybe convert it to a `HashSet` first which offers more efficient lookup.

Second, the logic for checking if the Event is interesting to us, and the logic for updating `busyTimes` is kind of mixed together when I think we could separate them a bit and reduce the level of nesting in the method. You could first check if any attendees are in our Set, and if not continue to the next Event, and if so then do the next piece of logic. Something like this:
```
Set<String> mandatoryAttendeeSet = new HashSet<>();
...
for (Event event : events) {
  boolean hasMandatoryAttendee =
    event.getAttendees().stream().filter(mandatoryAttendeeSet::contains).findAny().isPresent();

  if (!hasMandatoryAttendee) {
    continue;
  }

  ... build new busyTimes...
```

What do you think?

Reply:
I see, yes this division of logic makes sense. I have updated the code accordingly.